File: src/cpp/src/utils/no_metals/neighbor_radicals.cpp
#include "molgr/utils/no_metals/neighbor_radicals.h"

#include "molgr/stages/fresh.h"
#include "molgr/utils/conversions.h"
#include "molgr/utils/resonance.h"

#include "molgr/compat/openbabel_iter.h"

#include <openbabel/bond.h>

#include <algorithm>
#include <memory>
#include <optional>
#include <set>
#include <string>
#include <tuple>
#include <utility>
#include <variant>
#include <vector>

namespace
{
    constexpr std::size_t kMaxNeighborRadicalSeeds = 256;
    const std::string kChargeSeparationDiscrepancyKey =
        "neighbor_radical_charge_separation_discrepancy";
    using StateKey = std::pair<molgr::resonance::ResonanceStateKey, int>;
    using EnumerationKey = std::tuple<molgr::resonance::ResonanceStateKey, int, int>;

    StateKey BuildStateKey(const molgr::state::ReconstructionState &state)
    {
        return {molgr::resonance::BuildResonanceStateKey(state.Mol()), state.given_charge};
    }

    int ChargeSeparationDiscrepancy(const molgr::state::ReconstructionState &state)
    {
        const auto it = state.metadata.find(kChargeSeparationDiscrepancyKey);
        if (it != state.metadata.end())
        {
            if (const auto *value = std::get_if<int>(&it->second))
            {
                return *value;
            }
        }
        return 0;
    }

    bool ResolveNeighborRadicalPair(
        OpenBabel::OBMol &mol,
        int begin_idx,
        int end_idx,
        const std::string &mode,
        int positive_atom_idx)
    {
        auto *begin_atom = mol.GetAtom(begin_idx);
        auto *end_atom = mol.GetAtom(end_idx);
        if (begin_atom == nullptr || end_atom == nullptr)
        {
            return false;
        }
        const int spin_to_consume = std::min(
            begin_atom->GetSpinMultiplicity(),
            end_atom->GetSpinMultiplicity());
        if (spin_to_consume <= 0)
        {
            return false;
        }

        begin_atom->SetSpinMultiplicity(begin_atom->GetSpinMultiplicity() - spin_to_consume);
        end_atom->SetSpinMultiplicity(end_atom->GetSpinMultiplicity() - spin_to_consume);
        if (mode == "bond_order")
        {
            auto *bond = mol.GetBond(begin_idx, end_idx);
            if (bond == nullptr)
            {
                return false;
            }
            bond->SetBondOrder(bond->GetBondOrder() + spin_to_consume);
            molgr::reconstruct::AssignChargeRadicalForAtom(*begin_atom);
            molgr::reconstruct::AssignChargeRadicalForAtom(*end_atom);
            return true;
        }

        auto *positive_atom = positive_atom_idx == begin_idx ? begin_atom : end_atom;
        auto *negative_atom = positive_atom_idx == begin_idx ? end_atom : begin_atom;
        positive_atom->SetFormalCharge(positive_atom->GetFormalCharge() + spin_to_consume);
        negative_atom->SetFormalCharge(negative_atom->GetFormalCharge() - spin_to_consume);
        return true;
    }

    std::optional<molgr::state::ReconstructionState> ResolveSeed(
        const molgr::state::ReconstructionState &state,
        int begin_idx,
        int end_idx,
        const std::string &mode,
        int positive_atom_idx)
    {
        auto machine = molgr::state::OmolStateMachine::FromReconstructionState(state);
        auto next_mol = std::make_shared<OpenBabel::OBMol>(
            molgr::utils::CloneMolTopologyOnly(state.Mol()));
        machine = machine.Branch(std::nullopt, std::move(next_mol));
        const std::string phase = mode == "bond_order"
                                      ? "resolve_neighbor_radicals_by_bond_order"
                                      : "resolve_neighbor_radicals_by_charge_separation";
        const bool hit = machine.RunOmolStage(
            phase,
            ResolveNeighborRadicalPair,
            begin_idx,
            end_idx,
            mode,
            positive_atom_idx);
        if (!hit)
        {
            return std::nullopt;
        }

        std::string action;
        if (mode == "bond_order")
        {
            action = "bond_order:" + std::to_string(begin_idx) + "-" + std::to_string(end_idx);
            machine.metadata.erase("positive_atom_idx");
        }
        else
        {
            const int negative_atom_idx = positive_atom_idx == begin_idx ? end_idx : begin_idx;
            action = "charge_separation:" + std::to_string(positive_atom_idx) + "+" +
                     std::to_string(negative_atom_idx) + "-";
            machine.metadata["positive_atom_idx"] = positive_atom_idx;
        }
        const auto actions_it = machine.metadata.find("neighbor_radical_actions");
        if (actions_it != machine.metadata.end())
        {
            if (const auto *previous = std::get_if<std::string>(&actions_it->second))
            {
                action = *previous + ";" + action;
            }
        }
        machine.metadata["neighbor_radical_actions"] = action;
        machine.metadata["neighbor_radical_resolution"] = mode;
        int discrepancy = ChargeSeparationDiscrepancy(state);
        if (mode == "charge_separation")
        {
            ++discrepancy;
        }
        machine.metadata[kChargeSeparationDiscrepancyKey] = discrepancy;
        return machine.FreezeLike(state);
    }
}

namespace molgr
{
    namespace no_metals
    {
        namespace neighbor_radicals
        {
            std::vector<std::pair<int, int>> NeighborRadicalBondPairs(
                const OpenBabel::OBMol &mol)
            {
                std::set<std::pair<int, int>> unique;
                auto &mutable_mol = const_cast<OpenBabel::OBMol &>(mol);
                FOR_BONDS_OF_MOL(bond_iter, mutable_mol)
                {
                    auto *begin_atom = bond_iter->GetBeginAtom();
                    auto *end_atom = bond_iter->GetEndAtom();
                    if (begin_atom != nullptr && end_atom != nullptr &&
                        begin_atom->GetSpinMultiplicity() > 0 &&
                        end_atom->GetSpinMultiplicity() > 0)
                    {
                        unique.emplace(
                            std::min(begin_atom->GetIdx(), end_atom->GetIdx()),
                            std::max(begin_atom->GetIdx(), end_atom->GetIdx()));
                    }
                }
                return {unique.begin(), unique.end()};
            }

            std::vector<molgr::state::ReconstructionState> EnumerateNeighborRadicalSeeds(
                const molgr::state::ReconstructionState &state,
                std::optional<int> exact_discrepancy)
            {
                if (exact_discrepancy.has_value() && *exact_discrepancy < 0)
                {
                    return {};
                }
                if (NeighborRadicalBondPairs(state.Mol()).empty())
                {
                    auto machine = molgr::state::OmolStateMachine::FromReconstructionState(state);
                    machine.Annotate("neighbor_radicals_not_present");
                    machine.metadata["neighbor_radical_resolution"] = std::string("none");
                    machine.metadata[kChargeSeparationDiscrepancyKey] = 0;
                    if (exact_discrepancy.has_value() && *exact_discrepancy != 0)
                    {
                        return {};
                    }
                    return {machine.FreezeLike(state)};
                }

                std::vector<molgr::state::ReconstructionState> pending{state};
                std::vector<molgr::state::ReconstructionState> finished;
                std::set<EnumerationKey> expanded;
                std::set<StateKey> finished_keys;
                while (!pending.empty() && expanded.size() < kMaxNeighborRadicalSeeds)
                {
                    auto current = std::move(pending.back());
                    pending.pop_back();
                    const auto current_key = BuildStateKey(current);
                    const int current_discrepancy = ChargeSeparationDiscrepancy(current);
                    const int discrepancy_key = exact_discrepancy.has_value()
                                                    ? current_discrepancy
                                                    : 0;
                    const EnumerationKey enumeration_key{
                        current_key.first,
                        current_key.second,
                        discrepancy_key,
                    };
                    if (!expanded.insert(enumeration_key).second)
                    {
                        continue;
                    }
                    if (exact_discrepancy.has_value() &&
                        current_discrepancy > *exact_discrepancy)
                    {
                        continue;
                    }
                    const auto pairs = NeighborRadicalBondPairs(current.Mol());
                    if (pairs.empty())
                    {
                        if ((!exact_discrepancy.has_value() ||
                             current_discrepancy == *exact_discrepancy) &&
                            finished_keys.insert(current_key).second)
                        {
                            finished.push_back(std::move(current));
                        }
                        continue;
                    }

                    for (const auto &[begin_idx, end_idx] : pairs)
                    {
                        const std::vector<std::pair<std::string, int>> actions{
                            {"bond_order", 0},
                            {"charge_separation", begin_idx},
                            {"charge_separation", end_idx},
                        };
                        for (const auto &[mode, positive_atom_idx] : actions)
                        {
                            auto candidate = ResolveSeed(
                                current,
                                begin_idx,
                                end_idx,
                                mode,
                                positive_atom_idx);
                            if (candidate.has_value())
                            {
                                pending.push_back(std::move(*candidate));
                            }
                        }
                    }
                }

                std::sort(
                    finished.begin(),
                    finished.end(),
                    [](const auto &left, const auto &right)
                    {
                        const auto left_it = left.metadata.find("neighbor_radical_actions");
                        const auto right_it = right.metadata.find("neighbor_radical_actions");
                        const auto left_value = left_it == left.metadata.end()
                                                    ? std::string()
                                                    : std::get<std::string>(left_it->second);
                        const auto right_value = right_it == right.metadata.end()
                                                     ? std::string()
                                                     : std::get<std::string>(right_it->second);
                        return left_value < right_value;
                    });
                return finished;
            }
        }
    }
}

File: src/molgr/fallback/utils/no_metals/neighbor_radicals.py
"""Neighbor-radical seed enumeration for no-metal reconstruction."""

from __future__ import annotations

from typing import Literal, cast

from openbabel import openbabel as ob
from openbabel import pybel

from molgr.fallback.stages.fresh import assign_charge_radical_for_atom
from molgr.fallback.state import OmolStateMachine, ReconstructionState
from molgr.fallback.utils import resonance as resonance_utils
from molgr.fallback.utils.electrons import (
    get_unpaired_electron_count,
    set_unpaired_electron_count,
)


NeighborRadicalMode = Literal["bond_order", "charge_separation"]
_MAX_NEIGHBOR_RADICAL_SEEDS = 256
_CHARGE_SEPARATION_DISCREPANCY_KEY = "neighbor_radical_charge_separation_discrepancy"


def _clone_omol(omol: pybel.Molecule) -> pybel.Molecule:
    return pybel.Molecule(ob.OBMol(cast(ob.OBMol, omol.OBMol)))


def neighbor_radical_bond_pairs(omol: pybel.Molecule) -> list[tuple[int, int]]:
    """Return canonically oriented bonds whose endpoints both carry radicals."""

    pairs: list[tuple[int, int]] = []
    for bond in ob.OBMolBondIter(cast(ob.OBMol, omol.OBMol)):
        begin_atom = cast(ob.OBAtom, bond.GetBeginAtom())
        end_atom = cast(ob.OBAtom, bond.GetEndAtom())
        if (
            get_unpaired_electron_count(begin_atom) <= 0
            or get_unpaired_electron_count(end_atom) <= 0
        ):
            continue
        pairs.append(tuple(sorted((begin_atom.GetIdx(), end_atom.GetIdx()))))
    return sorted(set(pairs))


def _resolve_neighbor_radical_pair(
    omol: pybel.Molecule,
    begin_idx: int,
    end_idx: int,
    mode: NeighborRadicalMode,
    positive_atom_idx: int | None,
) -> tuple[pybel.Molecule, bool]:
    """Resolve adjacent real radicals by bonding or formal charge separation.

    Both modes consume ``min(unpaired_left, unpaired_right)`` real unpaired
    electrons from each endpoint. ``bond_order`` converts each pair into one bond
    increment and rebuilds endpoint labels. ``charge_separation`` converts the
    same count into ``+n/-n`` formal charges without using lone pairs or unresolved
    centers. The ``n > 1`` charge branch is a documented high-spin heuristic.
    Preconditions are validated before any electron field is changed.
    """

    obmol = cast(ob.OBMol, omol.OBMol)
    begin_atom = cast(ob.OBAtom, obmol.GetAtom(begin_idx))
    end_atom = cast(ob.OBAtom, obmol.GetAtom(end_idx))
    if begin_atom is None or end_atom is None:
        return omol, False

    spin_to_consume = min(
        get_unpaired_electron_count(begin_atom),
        get_unpaired_electron_count(end_atom),
    )
    if spin_to_consume <= 0:
        return omol, False

    bond = None
    if mode == "bond_order":
        bond = cast(ob.OBBond, obmol.GetBond(begin_idx, end_idx))
        if bond is None:
            return omol, False
    elif positive_atom_idx not in (begin_idx, end_idx):
        raise ValueError("positive_atom_idx must identify one endpoint of the radical bond")

    set_unpaired_electron_count(
        begin_atom, get_unpaired_electron_count(begin_atom) - spin_to_consume
    )
    set_unpaired_electron_count(end_atom, get_unpaired_electron_count(end_atom) - spin_to_consume)
    if mode == "bond_order":
        assert bond is not None
        bond.SetBondOrder(bond.GetBondOrder() + spin_to_consume)
        assign_charge_radical_for_atom(begin_atom)
        assign_charge_radical_for_atom(end_atom)
        return omol, True

    negative_atom = end_atom if positive_atom_idx == begin_idx else begin_atom
    positive_atom = begin_atom if positive_atom_idx == begin_idx else end_atom
    positive_atom.SetFormalCharge(positive_atom.GetFormalCharge() + spin_to_consume)
    negative_atom.SetFormalCharge(negative_atom.GetFormalCharge() - spin_to_consume)
    return omol, True


def _state_key(state: ReconstructionState) -> tuple[object, int]:
    return resonance_utils.build_resonance_state_key(state.omol), state.given_charge


def _enumeration_key(
    state: ReconstructionState,
    *,
    track_discrepancy: bool,
) -> tuple[object, ...]:
    state_key = _state_key(state)
    if not track_discrepancy:
        return state_key
    return (*state_key, int(state.metadata.get(_CHARGE_SEPARATION_DISCREPANCY_KEY, 0)))


def _resolved_seed(
    state: ReconstructionState,
    *,
    begin_idx: int,
    end_idx: int,
    mode: NeighborRadicalMode,
    positive_atom_idx: int | None,
) -> ReconstructionState | None:
    machine = OmolStateMachine.from_reconstruction_state(state).branch(
        None,
        omol=_clone_omol(state.omol),
    )
    phase = (
        "resolve_neighbor_radicals_by_bond_order"
        if mode == "bond_order"
        else "resolve_neighbor_radicals_by_charge_separation"
    )
    hit = machine.run_omol_stage(
        phase,
        _resolve_neighbor_radical_pair,
        begin_idx,
        end_idx,
        mode,
        positive_atom_idx,
    )
    if not hit:
        return None

    action = (
        f"bond_order:{begin_idx}-{end_idx}"
        if mode == "bond_order"
        else f"charge_separation:{positive_atom_idx}+"
        f"{end_idx if positive_atom_idx == begin_idx else begin_idx}-"
    )
    actions = tuple(state.metadata.get("neighbor_radical_actions", ())) + (action,)
    machine.metadata["neighbor_radical_actions"] = actions
    machine.metadata["neighbor_radical_resolution"] = mode
    discrepancy = int(state.metadata.get(_CHARGE_SEPARATION_DISCREPANCY_KEY, 0))
    if mode == "charge_separation":
        discrepancy += 1
    machine.metadata[_CHARGE_SEPARATION_DISCREPANCY_KEY] = discrepancy
    if positive_atom_idx is None:
        machine.metadata.pop("positive_atom_idx", None)
    else:
        machine.metadata["positive_atom_idx"] = positive_atom_idx
    return machine.freeze_like(state)


def enumerate_neighbor_radical_seeds(
    state: ReconstructionState,
    *,
    exact_discrepancy: int | None = None,
) -> list[ReconstructionState]:
    """Enumerate complete local resolutions of all adjacent radical pairs."""

    if exact_discrepancy is not None and exact_discrepancy < 0:
        return []

    if not neighbor_radical_bond_pairs(state.omol):
        machine = OmolStateMachine.from_reconstruction_state(state)
        machine.annotate("neighbor_radicals_not_present", neighbor_radical_resolution="none")
        candidate = machine.freeze_like(state)
        candidate.metadata[_CHARGE_SEPARATION_DISCREPANCY_KEY] = 0
        if exact_discrepancy not in (None, 0):
            return []
        return [candidate]

    pending = [state]
    finished: dict[tuple[object, int], ReconstructionState] = {}
    expanded: set[tuple[object, ...]] = set()
    while pending and len(expanded) < _MAX_NEIGHBOR_RADICAL_SEEDS:
        current = pending.pop()
        current_key = _state_key(current)
        enumeration_key = _enumeration_key(
            current,
            track_discrepancy=exact_discrepancy is not None,
        )
        if enumeration_key in expanded:
            continue
        expanded.add(enumeration_key)
        current_discrepancy = int(current.metadata.get(_CHARGE_SEPARATION_DISCREPANCY_KEY, 0))
        if exact_discrepancy is not None and current_discrepancy > exact_discrepancy:
            continue
        pairs = neighbor_radical_bond_pairs(current.omol)
        if not pairs:
            if exact_discrepancy is None or current_discrepancy == exact_discrepancy:
                finished.setdefault(current_key, current)
            continue

        for begin_idx, end_idx in pairs:
            actions = (
                ("bond_order", None),
                ("charge_separation", begin_idx),
                ("charge_separation", end_idx),
            )
            for mode, positive_atom_idx in actions:
                candidate = _resolved_seed(
                    current,
                    begin_idx=begin_idx,
                    end_idx=end_idx,
                    mode=cast(NeighborRadicalMode, mode),
                    positive_atom_idx=positive_atom_idx,
                )
                if candidate is not None:
                    pending.append(candidate)

    return sorted(
        finished.values(),
        key=lambda candidate: tuple(candidate.metadata.get("neighbor_radical_actions", ())),
    )


__all__ = [
    "enumerate_neighbor_radical_seeds",
    "neighbor_radical_bond_pairs",
]
